TOTP window tests are still buggy

## Description

the TOTP window test is still buggy :eyes: :eyes: 

## Steps to Reproduce

It's flaky. Just run the 2fa tests a bunch of times.

## Expected Behavior

They pass and window correctly.

## Actual Behavior

They don't. 

```
>       assert now < window_end
E       assert datetime.datetime(2018, 9, 29, 13, 50, 32, 750672) < datetime.datetime(2018, 9, 29, 13, 50, 32)
```
Please provide screenshots where appropriate.

## Comments

Suggestions to fix, any other relevant information.

--- a/securedrop/tests/test_2fa.py
+++ b/securedrop/tests/test_2fa.py
@@ -18,19 +18,25 @@ from utils.instrument import InstrumentedApp
 
 @contextmanager
 def totp_window():
-    # To ensure we have enough time during a single TOTP window to do the
-    # whole test, optionally sleep.
-    now = datetime.now()
-    num_secs_elapsed_in_time_window = now.second % 30
-    num_secs_left_in_time_window = 30 - num_secs_elapsed_in_time_window
+    '''To ensure we have enough time during a single TOTP window to do the
+       whole test, optionally sleep.
+    '''
 
-    if num_secs_left_in_time_window < 5:
-        time.sleep(30 - num_secs_elapsed_in_time_window)
-        now = datetime.now()
-        window_end = now.replace(microsecond=0) + timedelta(seconds=30)
-    else:
-        window_end = now.replace(microsecond=0) + \
-            timedelta(seconds=(30 - num_secs_left_in_time_window))
+    now = datetime.now()
+    # `or 30` to ensure we don't have a zero-length window
+    seconds_left_in_window = ((30 - now.second) % 30) or 30
+
+    window_end = now.replace(microsecond=0) + \
+        timedelta(seconds=seconds_left_in_window)
+    window_end_delta = window_end - now
+
+    # if we have less than 5 seconds left in this window, sleep to wait for
+    # the next window
+    if window_end_delta < timedelta(seconds=5):
+        sleep_time = window_end_delta.seconds + \
+            window_end_delta.microseconds / 1000000.0
+        time.sleep(sleep_time)
+        window_end = window_end + timedelta(seconds=30)
 
     yield
 

